### PR TITLE
Fix undefined behavior in SmallString.withUTF8

### PR DESCRIPTION
**Pull request by atrick**
Fix undefined behavior in SmallString.withUTF8

    withUTF8 currently vends a typed UInt8 pointer to the underlying
    SmallString. That pointer type differs from SmallString's
    representation. It should simply vend a raw pointer, which would be
    both type safe and convenient for UTF8 data. However, since this
    method is already @inlinable, I added calls to bindMemory to prevent
    the optimizer from reasoning about access to the typed pointer that we
    vend.

    rdar://67983613 (Undefinied behavior in SmallString.withUTF8 is miscompiled)

    Additional commentary:

    SmallString creates a situation where there are two types, the
    in-memory type, (UInt64, UInt64), vs. the element type,
    UInt8. `UnsafePointer<T>` specifies the in-memory type of the pointee,
    because that's how C works. If you want to specify an element type,
    not the in-memory type, then you need to use something other than
    UnsafePointer to view the memory. A trivial `BufferView<UInt8>` would
    be fine, although, frankly, I think UnsafeRawPointer is a perfectly
    good type on its own for UTF8 bytes.

    Unfortunately, a lot of the UTF8 helper code is ABI-exposed, so to
    work around this, we need to insert calls to bindMemory at strategic
    points to avoid undefined behavior. This is high-risk and can
    negatively affect performance. So far, I was able to resolve the
    regressions in our microbenchmarks just by tweaking the inliner.

### CODE REVIEW

**Comment by atrick** [comment]
Actually there is still one regression that I'll just file a separate bug for:
MapReduceString               39      46      +17.9%   **0.85x**

**Comment by atrick** [comment]
Incidentally, some of the BufferPointer code could be cleaner by relying on convenience methods that calculate capacity, but I think this code should stick to the most primitive operations to guarantee performance.

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
The current stdlib build does not appear to miscompile, but the miscompile is easily exposed by tweaking the optimizer in unrelated ways, and it *could* happen in user code since withUTF8 is inlinable.

**Comment by benrimmington** [comment]
#33693 fixes macOS test failure.

**Comment by benrimmington** [comment]
@swift-ci Please test macOS platform

**Comment by atrick** [comment]
@milseman do you have any concerns with this?

**Review by milseman — APPROVED** [review]
LGTM

File: stdlib/public/core/SmallString.swift [hunk 1]
```diff
@@ -197,26 +197,30 @@ extension _SmallString {
...
   ) rethrows -> Result {
+    let count = self.count
     var raw = self.zeroTerminatedRawCodeUnits
-    return try Swift.withUnsafeBytes(of: &raw) { rawBufPtr in
-      let ptr = rawBufPtr.baseAddress._unsafelyUnwrappedUnchecked
-        .assumingMemoryBound(to: UInt8.self)
-      return try f(UnsafeBufferPointer(start: ptr, count: self.count))
+    return try Swift.withUnsafeBytes(of: &raw) {
+      let rawPtr = $0.baseAddress._unsafelyUnwrappedUnchecked
+      // rebind the underlying (Uint64, UInt64) tuple to UInt8 for the
```
Comment (xwu) [inline stdlib/public/core/SmallString.swift:204]:
```suggestion
      // Rebind the underlying (UInt64, UInt64) tuple to UInt8 for the
```

File: stdlib/public/core/SmallString.swift [hunk 2]
```diff
@@ -197,26 +197,30 @@ extension _SmallString {
...
+      // duration of the closure. Accessing self after this rebind is undefined.
+      let ptr = rawPtr.bindMemory(to: UInt8.self, capacity: count)
+      defer {
+        // restore self's memory type
```
Comment (xwu) [inline stdlib/public/core/SmallString.swift:208]:
```suggestion
        // Restore the memory type of self._storage
```

File: stdlib/public/core/SmallString.swift [hunk 3]
```diff
@@ -273,7 +277,17 @@ extension _SmallString {
   ) rethrows {
     self.init()
     try self.withMutableCapacity {
-      return try initializer($0)
+      let capacity = $0.count
+      let rawPtr = $0.baseAddress._unsafelyUnwrappedUnchecked
+      // rebind the underlying (Uint64, UInt64) tuple to UInt8 for the
```
Comment (xwu) [inline stdlib/public/core/SmallString.swift:282]:
```suggestion
      // Rebind the underlying (UInt64, UInt64) tuple to UInt8 for the
```

File: stdlib/public/core/SmallString.swift [hunk 4]
```diff
@@ -273,7 +277,17 @@ extension _SmallString {
...
+      // duration of the closure. Accessing self after this rebind is undefined.
+      let ptr = rawPtr.bindMemory(to: UInt8.self, capacity: capacity)
+      defer {
+        // restore the memory type of self._storage
```
Comment (xwu) [inline stdlib/public/core/SmallString.swift:286]:
```suggestion
        // Restore the memory type of self._storage
```

**Review by xwu — COMMENTED** [review]
Nits :)

**Comment by atrick** [comment]
> Nits :)

Corrections have all been committed. Much appreciated.

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
I haven't changed these commits, but now there are large regressions.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by atrick** [comment]
I just hit another miscompile because of this after making an a small change to inlining, so I really need to merge it without investigating regressions. The performance regressions that popped up recently weren't originally a problem, so I'm sure they are fixable but will need a little time to investigate.

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test

**Comment by benrimmington** [comment]
> I haven't changed these commits, but now there are large regressions.

@atrick Could the previous regressions ([on 25 Sep][]) have been due to the creation of the `main` branch ([on 23 Sep][]), somehow affecting the baseline used by benchmarks?

[on 23 Sep]: <https://forums.swift.org/t/updating-branch-names/40412/4>
[on 25 Sep]: <https://github.com/apple/swift/pull/33698#issuecomment-698689809>

**Comment by atrick** [comment]
> > I haven't changed these commits, but now there are large regressions.
> 
> @atrick Could the previous regressions ([on 25 Sep](https://github.com/apple/swift/pull/33698#issuecomment-698689809)) have been due to the creation of the `main` branch ([on 23 Sep](https://forums.swift.org/t/updating-branch-names/40412/4)), somehow affecting the baseline used by benchmarks?

Possibly, I never saw a problem with local benchmarking.

**Pull request merged**